test/e2e/daemon_set.go

```diff
@@ -469,3 +566,64 @@ func checkDaemonPodsImage(c clientset.Interface, ns string, selector map[string]
 		return true, nil
 	}
 }
+
+func templateGenerationMatch(c clientset.Interface, ns string, selector map[string]string, templateGeneration string) func() (bool, error) {
+	return func() (bool, error) {
+		err := checkDaemonPodsTemplateGeneration(c, ns, selector, templateGeneration)
+		match := err == nil
+		return match, nil
+	}
+}
+
+func checkDaemonPodsTemplateGeneration(c clientset.Interface, ns string, label map[string]string, templateGeneration string) error {
+	pods := listDaemonPods(c, ns, label)
+	for _, pod := range pods.Items {
+		podTemplateGeneration := pod.Labels[extensions.DaemonSetTemplateGenerationKey]
+		if podTemplateGeneration != templateGeneration {
+			return fmt.Errorf("Expected pod %s/%s template generation %s, but got %s", pod.Namespace, pod.Name, templateGeneration, podTemplateGeneration)
+		}
+	}
+	return nil
+}
+
+func checkDaemonSetDeleted(f *framework.Framework, ns, name string) func() (bool, error) {
+	return func() (bool, error) {
+		_, err := f.ClientSet.Extensions().DaemonSets(ns).Get(name, metav1.GetOptions{})
+		if !apierrs.IsNotFound(err) {
+			return false, err
+		}
+		return true, nil
+	}
+}
+
+func getDaemonPodsLastCreationTime(c clientset.Interface, ns string, label map[string]string) metav1.Time {
+	sortedPods := getDaemonPodsSortedByCreationTime(c, ns, label)
+	return sortedPods[len(sortedPods)-1].ObjectMeta.CreationTimestamp
+}
+
+func getDaemonPodsFirstCreationTime(c clientset.Interface, ns string, label map[string]string) metav1.Time {
+	sortedPods := getDaemonPodsSortedByCreationTime(c, ns, label)
+	return sortedPods[0].ObjectMeta.CreationTimestamp
+}
+
+func getDaemonPodsSortedByCreationTime(c clientset.Interface, ns string, label map[string]string) []v1.Pod {
+	podList := listDaemonPods(c, ns, label)
+	pods := podList.Items
+	if len(pods) > 1 {
+		sort.Sort(podByCreationTimestamp(pods))
+	}
+	return pods
+}
+
+// podByCreationTimestamp sorts a list of DaemonSet pods by creation timestamp, using their names as a tie breaker.
+type podByCreationTimestamp []v1.Pod
```

Comment:
We have a `podByCreationTimestamp` in daemonset controller but it's `[]*v1.Pod` not `[]v1.Pod`